File: API/actuators/actuators.cpp
#include <stdio.h>

class actuator {

	public:

		actuator(const char *name, const char *pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}

		// This method depends on quite a bit. Depending on the data communication
		// type and the sensor, it will use different libraries and return different 
		// values.
		bool toggle();
};

class EC : public actuator {

	const char *units	        = "micro-S-cm-l";
	const char *data_comm	        = "I2C";
	const float sensor_range[2]     = {3, 3000};
	const float required_resolution = 1;

	public:

		EC(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class pH : public actuator {

	const char *units	        = "pH";
	const char *data_comm	        = "I2C";
	const float sensor_range[2]     = {2, 12};
	const float required_resolution = .2;

	public:

		pH(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class temperature : public actuator {

	const char *units	        = "C";
	const char *data_comm	        = "one-wire";
	const float sensor_range[2]     = {0, 100};
	const float required_resolution = 1;

	public:

		temperature(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class moisture : public actuator {

	const char *units	        = "%";
	const char *data_comm	        = "ADC-I2C";
	const float sensor_range[2]     = {0, 50};
	const float required_resolution = 1;

	public:

		moisture(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class DO_probe : public actuator {

	const char *units	        = "mg/L";
	const char *data_comm	        = "I2C";
	const float sensor_range[2]     = {0, 15};
	const float required_resolution = .1;

	public:

		DO_probe(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class liquid_level : public actuator {

	const char *units	        = "cm";
	const char *data_comm	        = "ADC-I2C";
	const float sensor_range[2]     = {0, 40.5};
	const float required_resolution = 1.25;

	public:

		liquid_level(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class flow_meter : public actuator {

	const char *units	        = "gpm";
	const char *data_comm	        = "UART";
	const float sensor_range[2]     = {.2, 2};
	const float required_resolution = .05;

	public:

		flow_meter(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class RH_temp : public actuator {

	const char *units	        = "(C, %)";
	const char *data_comm	        = "one-wire";
	const float sensor_range[2]     = {5, 99};
	const float required_resolution = 1;

	public:

		RH_temp(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class total_pressure : public actuator {

	const char *units	        = "kPA";
	const char *data_comm	        = "I2C";
	const float sensor_range[2]     = {30, 110};
	const float required_resolution = 1;

	public:

		total_pressure(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class oxygen : public actuator {

	const char *units	        = "%";
	const char *data_comm	        = "ADC-I2C";
	const float sensor_range[2]     = {0, 100};
	const float required_resolution = 1;

	public:

		oxygen(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class CO2 : public actuator {

	const char *units	        = "ppm";
	const char *data_comm	        = "UART";
	const float sensor_range[2]     = {0, 2000};
	const float required_resolution = 100;

	public:

		CO2(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class light_PAR : public actuator {

	const char *units	        = "micro-mol per meters-squared seconds";
	const char *data_comm	        = "ADC-I2C";
	const float sensor_range[2]     = {0, 2000};
	const float required_resolution = 2;

	public:

		light_PAR(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

class camera : public actuator {

	const char *units	    = "RGB";
	const char *data_comm	    = "USB";
	const float sensor_range[2] = {1, 255};
	//float required_resolution = "1k0x1k";

	public:

		camera(const char *name, const char *pin) 
		: sensor(name, pin) {

			// these are determined at construction
			name = name;
			pin  = pin;
		}
};

int main()
{
	actuator F1 = actuator("S101", "0x66");
	actuator F2 = actuator("S102", "0x65");
	actuator F3 = actuator("S103", NULL);
	actuator F4 = actuator("S104", "0x61");
	actuator F5 = actuator("S105", NULL);

	actuator M01 = actuator("S106", NULL);
	actuator M02 = actuator("S106", NULL);
	actuator M03 = actuator("S106", NULL);

	actuator M05 = actuator("S106", NULL);
	actuator M07 = actuator("S107", NULL);
	actuator M08 = actuator("S108", NULL);
	actuator M09 = actuator("S109", NULL);
	actuator M10 = actuator("S110", NULL);
	actuator M11 = actuator("S111", NULL);
	actuator M12 = actuator("S112", NULL);
	
	actuator M15 = actuator("S112", NULL);
	actuator M16 = actuator("S112", NULL);
	actuator M17 = actuator("S112", NULL);
	actuator M18 = actuator("S112", NULL);
	actuator M19 = actuator("S112", NULL);

	actuator P01 = actuator("S201", NULL);
	actuator P02 = actuator("S202", NULL);
	actuator P03 = actuator("S203", NULL);
	actuator P04 = actuator("S204", NULL);
	actuator P05 = actuator("S205", NULL);
	actuator P06 = actuator("S206", NULL);
	actuator P07 = actuator("S206", NULL);
	actuator P08 = actuator("S208", NULL);
	actuator P09 = actuator("S209", NULL);
	actuator P10 = actuator("S210", NULL);
	actuator P11 = actuator("S211", NULL);
	actuator P12 = actuator("S211", NULL);

	actuator R01 = actuator("S301", NULL);
	actuator R02 = actuator("S302", "P8_9");

	actuator T03 = actuator("S303", "0x77");
	actuator T03 = actuator("S303", "0x77");
	actuator T03 = actuator("S303", "0x77");
	actuator T04 = actuator("S304", NULL);
	actuator T05 = actuator("S305", NULL);
	actuator T06 = actuator("S306", NULL);
	actuator T07 = actuator("S307", NULL);
	actuator T08 = actuator("S307", NULL);
	actuator T09 = actuator("S307", NULL);

	actuator V402 = actuator("S402", NULL);
	actuator V403 = actuator("S403", NULL);
	actuator V404 = actuator("S403", NULL);
	actuator V405 = actuator("S403", NULL);

	actuator V407 = actuator("S403", NULL);

	actuator Z01 = actuator("S201", NULL);
	actuator Z02 = actuator("S202", NULL);
	actuator Z03 = actuator("S203", NULL);
	actuator Z04 = actuator("S204", NULL);
	actuator Z05 = actuator("S205", NULL);
	actuator Z06 = actuator("S206", NULL);
	actuator Z07 = actuator("S206", NULL);
	actuator Z08 = actuator("S208", NULL);
	actuator Z09 = actuator("S209", NULL);
	actuator Z10 = actuator("S210", NULL);
	actuator Z11 = actuator("S211", NULL);
	actuator Z12 = actuator("S211", NULL);
	actuator Z13 = actuator("S211", NULL);
}

File: control/API/actuators.py
class actuator:

	def __init__(self, name, pin):

		# these are determined at construction
		name = name;
		pin  = pin;

class GPIO_actuator(actuator): 

	BBB_GPIO = {"P9_12": 60,
		    "P9_25": 117,
		    "P9_28": 113,
		    "P9_19": 13,
		    "P9_20": 12,
		    "P8_39": 76,
		    "P8_40": 77,
		    "P8_41": 74,
		    "P8_42": 75,
		    "P8_43": 72,
		    "P8_44": 73,
		    "P9_14": 50,
		    "P9_15": 48,
		    "P9_23": 49,
		    "P9_29": 111,
		    "P9_30": 112,
		    "P9_31": 110,
		    "P9_41": 116,
		    "P9_42": 7}

	def toggle(self): pass

class PWM_actuator(actuator): pass

F1 = actuator("F1")
F2 = actuator("F2")
F3 = actuator("F3")
F4 = actuator("F4")
F5 = actuator("F5")

M01 = actuator("M01")
M02 = actuator("M02")
M03 = actuator("M03")

M05 = actuator("M05")
M06 = actuator("M06")
M07 = actuator("M07")
M08 = actuator("M08")
M09 = actuator("M09")
M10 = actuator("M10")
M11 = actuator("M11")
M12 = actuator("M12")

M15 = actuator("M15")
M16 = actuator("M16")
M17 = actuator("M17")
M18 = actuator("M18")
M19 = actuator("M19")

P01 = actuator("P01")
P02 = actuator("P02")
P03 = actuator("P03")
P04 = actuator("P04")
P05 = actuator("P05")
P06 = actuator("P06")
P07 = actuator("P06")
P08 = actuator("P08")
P09 = actuator("P09")
P10 = actuator("P10")
P11 = actuator("P11")
P12 = actuator("P12")

R01 = actuator("R01")
R02 = actuator("R02")

T01 = actuator("T01")
T02 = actuator("T02")
T03 = actuator("T03")
T04 = actuator("T04")
T05 = actuator("T05")
T06 = actuator("T06")
T07 = actuator("T07")
T08 = actuator("T08")
T09 = actuator("T09")

V02 = actuator("V02")
V03 = actuator("V03")
V04 = actuator("V04")
V05 = actuator("V05")

V07 = actuator("V07")

Z01 = actuator("Z01")
Z02 = actuator("Z02")
Z03 = actuator("Z03")
Z04 = actuator("Z04")
Z05 = actuator("Z05")
Z06 = actuator("Z06")
Z07 = actuator("Z06")
Z08 = actuator("Z08")
Z09 = actuator("Z09")
Z10 = actuator("Z10")
Z11 = actuator("Z11")
Z12 = actuator("Z12")
Z13 = actuator("Z13")

actuator_suite = (F1,  F2,  F3,  F4,  F5,
		  M01, M02, M03,      M05, M06, M07, M08, M09, M10, M11, M12,           M15, M16, M17, M18, M19,
		  P01, P02, P03, P04, P05, P06, P07, P08, P09, P10, P11, P12,
		  R01, R02,
		  T01, T02, T03, T04, T05, T06, T07, T08, T09,
		       V02, V03, V04, V05,      V07, 
		  Z01, Z02, Z03, Z04, Z05, Z06, Z07, Z08, Z09, Z10, Z11, Z12, Z13)
